File: python/basics/booleans.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
  booleans.py: Exploring boolean literals in Python

  This is an exploration on the usage of numeric literals in Python
"""
# -------------------------------------------------------------------
# booleans.py: Exploring boolean literals in python
#
# Copyright (C) 2024 Sumanth Vepa.
#
# This program is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License a
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <https://www.gnu.org/licenses/>.
# -------------------------------------------------------------------


# There are two boolean literals in Python: True and False.
def explore_boolean_literals():
  """
  Explore boolean literals in Python
  """
  # This is a boolean literal. The value True.
  true_literal = True
  print(true_literal)
  # This is a boolean literal. The value False.
  false_literal = False
  print(false_literal)

  # You can also use the bool() function to convert other values to
  # boolean values.
  print(bool(0))  # False
  print(bool(1))  # True
  print(bool(42))  # True
  print(bool(-1))  # True
  print(bool(0.0))  # False
  print(bool(0.1))  # True
  print(bool(0.0+0.0j))  # False
  print(bool(0.0+1.0j))  # True

  # You can also use the bool() function to convert strings to boolean
  # values.
  print(bool(''))  # False
  print(bool(' '))  # True
  print(bool('Hello, World!'))  # True

  # You can also use the bool() function to convert lists to boolean
  # values.
  print(bool([]))  # False
  print(bool([1, 2, 3]))  # True

  # You can also use the bool() function to convert tuples to boolean
  # values.
  print(bool(()))  # False
  print(bool((1, 2, 3)))  # True

  # You can also use the bool() function to convert dictionaries to
  # boolean values.
  print(bool({}))  # False
  print(bool({'a': 1, 'b': 2, 'c': 3}))  # True

  # You can also use the bool() function to convert sets to boolean
  # values.
  print(bool(set()))  # False
  print(bool({1, 2, 3}))  # True

  # You can also use the bool() function to convert None to boolean
  # values.
  print(bool(None))  # False

  # You can also use the bool() function to convert functions to
  # boolean values.
  def my_function():
    return 42
  print(bool(my_function))  # True

  # You can also use the bool
  # function to convert classes to boolean values.
  # pylint: disable=too-few-public-methods
  class MyClass:
    """ A simple class"""
  print(bool(MyClass))  # True

File: swift/basics/basics/booleans.swift
//-*- coding: utf-8 -*-
/**
  booleans.swift: Explore booleans in Swift
 
  This is an exploration of booleans in Swift
*/
/* -------------------------------------------------------------------
 * booleans.swift: Explore booleans in Swift.
 *
 * Copyright (C) 2024 Sumanth Vepa.
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License a
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see
 * <https://www.gnu.org/licenses/>.
 *-----------------------------------------------------------------*/
import Foundation


func exploreBooleans() {

  // Booleans in Swift are declared with the Bool type.
  // They can have two values: true and false.
  // The type annotation is Bool. As usual, if you
  // initialize the variable, the type is inferred. (Note: VSCode
  // will show the inferred type inline in gray. This is not part
  // of the text file itself.)
  let isTrue = true
  let isFalse = false

  print("isTrue = ", isTrue)
  print("type(of: isTrue) = ", type(of: isTrue))
  print("isFalse = ", isFalse)
  print("type(of: isFalse) = ", type(of: isFalse))

  // Booleans can be used in conditional statements.
  let number: UInt32 = arc4random_uniform(10)
  let condition = number >= 5
  if condition {
    print("condition is true")
  } else {
    print("condition is false")
  }

  let not_condition = !condition
  if not_condition {
    print("not_condition is true")
  } else {
    print("not_condition is false")
  }

  // Booleans can be used in conditional expressions.
  let condition2: Bool = number > 2
  let result = condition2 ? "number is > 2" : "number <= 2"
  print(result)

  // You can explicitly specify the type of a boolean.
  var condition3: Bool
  condition3 = number < 5
  if condition3 {
    print("number < 5")
  } else {
    print("number > 5")
  }
}
